### PR TITLE
Column mapping

### PR DESCRIPTION
**Pull request by binglihub**

### CODE REVIEW

File: src/main/scala/net/snowflake/spark/snowflake/Parameters.scala [hunk 1]
```diff
@@ -67,6 +67,10 @@ object Parameters {
   val PARAM_AWS_ACCESS_KEY     = knownParam("awsaccesskey")
   val PARAM_USE_STAGING_TABLE  = knownParam("usestagingtable")
   val PARAM_EXTRA_COPY_OPTIONS = knownParam("extracopyoptions")
+  val PARAM_AUTO_PUSHDOWN      = knownParam("autopushdown")
+  val PARAM_COLUMN_MAP         = knownParam("columnmap")
```
Comment (etduwx) [inline src/main/scala/net/snowflake/spark/snowflake/Parameters.scala]: formatting

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeWriter.scala [hunk 2]
```diff
@@ -288,10 +288,52 @@ private[snowflake] class SnowflakeWriter(
...
+  private def copySql(data: DataFrame,
+                      saveMode: SaveMode,
                       params: MergedParameters,
                       filesToCopy: (String, String),
                       tempStage: Option[String]): String = {
+
+    if (saveMode != SaveMode.Append && params.columnMap.isDefined)
+      throw new UnsupportedOperationException("The column mapping only works in append mode.")
+
+    def getMappingToString(list: Option[List[(Int, String)]]): String = {
```
Comment (etduwx) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeWriter.scala]: Where are we checking compatibility with the target table? What if the columns don't exist in the target table?

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeConnectorUtils.scala [hunk 3]
```diff
@@ -12,6 +12,22 @@ object SnowflakeConnectorUtils {
 
   @transient lazy val log = LoggerFactory.getLogger(getClass.getName)
 
+
+  /**
+    * Check Spark version, if Spark version matches SUPPORT_SPARK_VERSION enable PushDown, otherwise disable it.
+    */
+  private val SUPPORT_SPARK_VERSION = "2.2"
+
+  def checkSparkVersion(session: SparkSession): Boolean =
```
Comment (etduwx) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeConnectorUtils.scala]: Change this to "checkVersionAndEnablePushdown"

File: src/it/scala/net/snowflake/spark/snowflake/ColumnMappingIntegrationSuite.scala [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    checkAnswer(loadedDf, expectedAnswer)
+  }
+
+  test("Test columnMapping") {
+
+    jdbcUpdate(s"create or replace table $dbtable (ONE int, TWO int, THREE int, FOUR int, FIVE int)")
+
+
+    df.write.format(SNOWFLAKE_SOURCE_NAME).options(connectorOptionsNoTable)
+      .option("dbtable", dbtable)
```
Comment (etduwx) [inline src/it/scala/net/snowflake/spark/snowflake/ColumnMappingIntegrationSuite.scala]: Why does this throw an exception?

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeWriter.scala [hunk 5]
```diff
@@ -288,10 +288,52 @@ private[snowflake] class SnowflakeWriter(
...
+      if (list.isEmpty || list.get.isEmpty) ""
+      else {
+        val map = list.get
+        val buffer: mutable.StringBuilder = new mutable.StringBuilder()
+        buffer.append("(")
+        buffer.append(map.head._2)
+        (1 until map.size).foreach(i => {
+          buffer.append(", ")
+          buffer.append(map(i)._2)
```
Comment (etduwx) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeWriter.scala]: Let's ensure proper escaping of these identifiers. See: quotedName() and its usage in SnowflakeJDBCWrapper.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeConnectorUtils.scala (hunk 3 above)
Comment (binglihub) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeConnectorUtils.scala]: done

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeWriter.scala (hunk 5 above)
Comment (binglihub) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeWriter.scala]: done

File: src/it/scala/net/snowflake/spark/snowflake/ColumnMappingIntegrationSuite.scala (hunk 4 above)
Comment (binglihub) [reply src/it/scala/net/snowflake/spark/snowflake/ColumnMappingIntegrationSuite.scala]: I added some new comments beside these test cases in the new commit

**Pull request merged**